File: patient-ios-app/MediHubPatient/Core/Models/MedicalHistoryModels.swift
import Foundation

// MARK: - My Medical History (matches the web's my-medical-history)

// Four read-only surfaces under /me/patient/*. The backend builds them for
// staff first, so each DTO carries far more than the patient page shows; only
// the fields the web reads are declared here, and everything but the id is
// optional so an omitted or null field never fails the decode.

/// PatientDiagnosisSummaryDTO — problems and diagnoses merged, newest first.
struct HistoryDiagnosis: Decodable, Identifiable {
    let id: String
    let description: String?
    let icdCode: String?
    let status: String?
    /// "yyyy-MM-dd'T'HH:mm:ssXXX"
    let diagnosedAt: String?
    let diagnosedByName: String?
}

/// PatientSurgicalHistoryResponseDTO (NON_NULL: absent fields are simply missing).
struct HistorySurgery: Decodable, Identifiable {
    let id: String
    let hospitalName: String?
    let procedureCode: String?
    let procedureDisplay: String?
    /// "yyyy-MM-dd"
    let procedureDate: String?
    let outcome: String?
    let performedBy: String?
    let location: String?
    let notes: String?
}

/// FamilyHistoryResponseDTO, the part the web lists.
struct HistoryFamilyEntry: Decodable, Identifiable {
    let id: String
    let relationship: String?
    let relativeName: String?
    let relativeGender: String?
    let relativeLiving: Bool?
    let relativeAge: Int?
    let relativeAgeAtDeath: Int?
    let causeOfDeath: String?
    let conditionCode: String?
    let conditionDisplay: String?
    let conditionCategory: String?
    let ageAtOnset: Int?
    let severity: String?
    let notes: String?
    let active: Bool?
}

/// SocialHistoryResponseDTO — the latest active record, or `data: null`
/// when none was ever recorded (the web shows "no social history" then).
struct HistorySocial: Decodable, Identifiable {
    let id: String
    let tobaccoUse: Bool?
    let tobaccoType: String?
    let tobaccoPacksPerDay: Double?
    let tobaccoYearsUsed: Int?
    /// "yyyy-MM-dd"
    let tobaccoQuitDate: String?
    let tobaccoNotes: String?
    let alcoholUse: Bool?
    let alcoholFrequency: String?
    let alcoholDrinksPerWeek: Int?
    let alcoholBingeDrinking: Bool?
    let alcoholNotes: String?
    let recreationalDrugUse: Bool?
    let exerciseFrequency: String?
    let exerciseType: String?
    let occupation: String?
    let employmentStatus: String?
    let maritalStatus: String?
    let active: Bool?
}

File: patient-android-app/app/src/main/java/com/bitnesttechs/hms/patient/core/models/MedicalHistoryModels.kt
package com.bitnesttechs.hms.patient.core.models

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass

/**
 * GET /me/patient/medical-history (PatientDiagnosisSummaryDTO, wrapped).
 * Problem-list entries and diagnoses merged newest first; read-only for the
 * patient, exactly as the web's My Medical History.
 */
@JsonClass(generateAdapter = true)
data class PatientDiagnosisSummary(
    @Json(name = "id") val id: String,
    @Json(name = "description") val description: String? = null,
    @Json(name = "icdCode") val icdCode: String? = null,
    @Json(name = "status") val status: String? = null,
    /** "yyyy-MM-dd'T'HH:mm:ssXXX" */
    @Json(name = "diagnosedAt") val diagnosedAt: String? = null,
    @Json(name = "diagnosedByName") val diagnosedByName: String? = null
)

/** GET /me/patient/surgical-history (PatientSurgicalHistoryResponseDTO, wrapped, NON_NULL fields only). */
@JsonClass(generateAdapter = true)
data class SurgicalHistoryEntry(
    @Json(name = "id") val id: String,
    @Json(name = "patientId") val patientId: String? = null,
    @Json(name = "hospitalId") val hospitalId: String? = null,
    @Json(name = "hospitalName") val hospitalName: String? = null,
    @Json(name = "procedureCode") val procedureCode: String? = null,
    @Json(name = "procedureDisplay") val procedureDisplay: String? = null,
    /** "yyyy-MM-dd" */
    @Json(name = "procedureDate") val procedureDate: String? = null,
    @Json(name = "outcome") val outcome: String? = null,
    @Json(name = "performedBy") val performedBy: String? = null,
    @Json(name = "location") val location: String? = null,
    @Json(name = "notes") val notes: String? = null
)

/** GET /me/patient/family-history (FamilyHistoryResponseDTO, wrapped). Only the fields the web shows are modelled. */
@JsonClass(generateAdapter = true)
data class FamilyHistoryEntry(
    @Json(name = "id") val id: String,
    @Json(name = "relationship") val relationship: String? = null,
    @Json(name = "relativeName") val relativeName: String? = null,
    @Json(name = "relativeGender") val relativeGender: String? = null,
    @Json(name = "relativeLiving") val relativeLiving: Boolean? = null,
    @Json(name = "relativeAge") val relativeAge: Int? = null,
    @Json(name = "relativeAgeAtDeath") val relativeAgeAtDeath: Int? = null,
    @Json(name = "causeOfDeath") val causeOfDeath: String? = null,
    @Json(name = "conditionCode") val conditionCode: String? = null,
    @Json(name = "conditionDisplay") val conditionDisplay: String? = null,
    @Json(name = "conditionCategory") val conditionCategory: String? = null,
    @Json(name = "ageAtOnset") val ageAtOnset: Int? = null,
    @Json(name = "severity") val severity: String? = null,
    @Json(name = "notes") val notes: String? = null,
    @Json(name = "active") val active: Boolean? = null
)

/**
 * GET /me/patient/social-history (SocialHistoryResponseDTO, wrapped; `data`
 * is null when nothing active is on record). Only the fields the web shows.
 */
@JsonClass(generateAdapter = true)
data class SocialHistory(
    @Json(name = "id") val id: String? = null,
    @Json(name = "tobaccoUse") val tobaccoUse: Boolean? = null,
    @Json(name = "tobaccoType") val tobaccoType: String? = null,
    @Json(name = "tobaccoPacksPerDay") val tobaccoPacksPerDay: Double? = null,
    @Json(name = "tobaccoYearsUsed") val tobaccoYearsUsed: Int? = null,
    /** "yyyy-MM-dd" */
    @Json(name = "tobaccoQuitDate") val tobaccoQuitDate: String? = null,
    @Json(name = "tobaccoNotes") val tobaccoNotes: String? = null,
    @Json(name = "alcoholUse") val alcoholUse: Boolean? = null,
    @Json(name = "alcoholFrequency") val alcoholFrequency: String? = null,
    @Json(name = "alcoholDrinksPerWeek") val alcoholDrinksPerWeek: Int? = null,
    @Json(name = "alcoholBingeDrinking") val alcoholBingeDrinking: Boolean? = null,
    @Json(name = "alcoholNotes") val alcoholNotes: String? = null,
    @Json(name = "recreationalDrugUse") val recreationalDrugUse: Boolean? = null,
    @Json(name = "exerciseFrequency") val exerciseFrequency: String? = null,
    @Json(name = "exerciseType") val exerciseType: String? = null,
    @Json(name = "occupation") val occupation: String? = null,
    @Json(name = "employmentStatus") val employmentStatus: String? = null,
    @Json(name = "maritalStatus") val maritalStatus: String? = null,
    @Json(name = "active") val active: Boolean? = null
)
